client/configuration.py

```diff
@@ -354,11 +354,16 @@ def _read(self, path: str) -> None:
                 configuration_directory = os.path.dirname(path)
                 if configuration_directory:
                     self.source_directories = [
-                        os.path.join(configuration_directory, directory)
+                        # TODO: Does it make sense to expanduser
+                        # when a configuration directory is present?
+                        os.path.join(configuration_directory, os.path.expanduser(directory))
```

Comment:
Since the directory is forced to be a relative path here, it seems unlikely that expanding `~` would change anything here. As a side note, checking whether the dirname of the path is not None seems a bit finicky. I suppose the proper thing to do here would be to detect whether the source directory is an absolute path and avoid joining it with the configuration directory in the first place, but that's definitely outside the scope of this PR.

Reply:
That's more or less what I was thinking, thanks. Also agree with the side note, but I'll avoid that in this PR for scope and test complexity while I'm still unfamiliar 🙂